### PR TITLE
[Feature] Support ball_query with cambricon MLU backend and mlu-ops library.

### PR DESCRIPTION
**Pull request by DanieeelLiu**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

The motivation of the PR is to support running ball_query with Cambricon MLU backend.
 
It includes five parts:
 
    1. Add mlu_common_desc.cpp  as mlu-ops adaptor.
    2. Add mlu_common_desc.h as part of mlu-ops adaptor.
    3. Add ball_query_mlu.cpp to support launching mlu-ops operator in PyTorch
    4. Refactor test_ball_query.py to support testing Ball_query_forward with MLU backend.
    5. Refactor setup.py to support compile mmcv project with mlu-ops to support MLU backend.
 
## Modification
 
- mlu-ops adaption
  Add cpp code of mlu-ops adaption in directory mmcv/ops/csrc/pytorch/mlu/mlu_common_desc.cpp and mmcv/ops/csrc/pytorch/mlu/mlu_common_desc.h.
- PyTorch adaptation
  Adapt ball_query_forward for PyTorch in mmcv/ops/csrc/pytorch/mlu/ball_query_mlu.cpp.
- Unit test
  Support test Ball_query with various backends in tests/test_ops/test_ball_query.py.
- Compilation
  Support compile mmcv with mlu-ops to support MLU backend in setup.py.
Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

**Before PR**:

- [x] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) to create this PR.
- [x] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) are used to fix the potential lint issues.
- [ ] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with some of those projects, like MMDet or MMCls.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmcv/ops/csrc/pytorch/mlu/mlu_common_desc.cpp [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+  if (mapping_type.find(std::string(data_type.name())) != mapping_type.end()) {
+    return mapping_type.find(std::string(data_type.name()))->second;
+  }
+  return MLUOP_DTYPE_INVALID;
+}
+
+void MluOpTensorDescriptor::set(Tensor t) {
+  mluOpDataType_t data_type = getMluOpDataType(t.dtype());
+  mluOpTensorLayout_t layout =
+      MLUOP_LAYOUT_ARRAY;  // XXX Improve it in the future;
```
Comment (Qiza-lyhm) [inline mmcv/ops/csrc/pytorch/mlu/mlu_common_desc.cpp:43]: check the layout info here. Might here shall be NHWC? Or just the same as the tensor stride info?

File: mmcv/ops/csrc/pytorch/mlu/ball_query_mlu.cpp [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  MluOpTensorDescriptor new_xyz_desc, xyz_desc, idx_desc;
+  new_xyz_desc.set(new_xyz);
+  xyz_desc.set(xyz);
+  idx_desc.set(idx);
+
+  auto new_xyz_impl = torch_mlu::getMluTensorImpl(new_xyz);
+  auto xyz_impl = torch_mlu::getMluTensorImpl(xyz);
+  auto idx_impl = torch_mlu::getMluTensorImpl(idx);
+  auto new_xyz_ptr = new_xyz_impl->cnnlMalloc();
+  auto xyz_ptr = xyz_impl->cnnlMalloc();
```
Comment (Qiza-lyhm) [inline mmcv/ops/csrc/pytorch/mlu/ball_query_mlu.cpp:29]: 请问这里是否需要对输入tensor做contiguous操作？还是已经在python端完成了？

File: mmcv/ops/csrc/pytorch/mlu/mlu_common_desc.cpp (hunk 1 above)
Comment (DanieeelLiu) [reply mmcv/ops/csrc/pytorch/mlu/mlu_common_desc.cpp]: Fixed. It will set layout according to tensor's layout.

Comment (Qiza-lyhm) [reply mmcv/ops/csrc/pytorch/mlu/mlu_common_desc.cpp]: okay

**Review by Qiza-lyhm — APPROVED** [review]

File: setup.py [hunk 3]
```diff
@@ -300,16 +301,54 @@ def get_extensions():
...
             define_macros += [('MMCV_WITH_MLU', None)]
             mlu_args = os.getenv('MMCV_MLU_ARGS')
-            extra_compile_args['cncc'] = [mlu_args] if mlu_args else []
+            mluops_includes = []
+            mluops_includes.append('-I' +
+                                   os.path.abspath('./mlu-ops/bangc-ops'))
+            mluops_includes.append(
+                '-I' + os.path.abspath('./mlu-ops/bangc-ops/kernels'))
+            extra_compile_args['cncc'] = [mlu_args] + \
+                mluops_includes if mlu_args else mluops_includes
```
Comment (Qiza-lyhm) [inline setup.py:335]: 为了避免mmcv中的mluops与其他库中的mluops发生符号冲突，此处建议添加如下改动：
extra_compile_args['cncc'] += '-fno-gnu-unique'
extra_compile_args['cxx'] += '-fno-gnu-unique'

Comment (DanieeelLiu) [reply setup.py:335]: Fixed

Comment (Qiza-lyhm) [reply setup.py:335]: nice

File: setup.py [hunk 4]
```diff
@@ -300,16 +301,55 @@ def get_extensions():
...
+                    'mlu-ops/releases/latest'
+                response = requests.get(url=mluops_url, allow_redirects=False)
+                download_url = response.headers.get('Location').replace(
+                    'releases/tag', 'archive/refs/tags') + '.zip'
+                req = requests.get(download_url)
+                with open('./mlu-ops-latest.zip', 'wb') as f:
+                    try:
+                        f.write(req.content)
+                    except Exception:
+                        raise ImportError('downlowad mlu-ops fail')
```
Comment (grimoire) [inline setup.py:316]: Typo.

Comment (DanieeelLiu) [reply setup.py]: Fixed

File: tests/test_ops/test_ball_query.py [hunk 5]
```diff
@@ -3,55 +3,60 @@
...
+        'cuda',
+        marks=pytest.mark.skipif(
+            not IS_CUDA_AVAILABLE, reason='requires CUDA support')),
+    pytest.param(
+        'mlu',
+        marks=pytest.mark.skipif(
+            not IS_MLU_AVAILABLE, reason='requires MLU support'))
+])
+def test_ball_query(device):
+    print(device)
```
Comment (grimoire) [inline tests/test_ops/test_ball_query.py:20]: Please remove print

**Comment by grimoire** [comment]
Test failed with `undefined symbol: __bangPopCallConfiguration`. Do I need to add some libraries to the environment?

File: tests/test_ops/test_ball_query.py (hunk 5 above)
Comment (DanieeelLiu) [reply tests/test_ops/test_ball_query.py]: Fixed

**Review by grimoire — APPROVED** [review]
LGTM

File: setup.py [hunk 6]
```diff
@@ -300,16 +301,72 @@ def get_extensions():
...
+                    response = requests.get(
+                        url=mluops_url, allow_redirects=False)
+                    download_url = response.headers.get('Location').replace(
+                        'releases/tag', 'archive/refs/tags') + '.zip'
+                    req = requests.get(download_url)
+                    with open('./mlu-ops-latest.zip', 'wb') as f:
+                        try:
+                            f.write(req.content)
+                        except Exception:
+                            raise ImportError('download mlu-ops fail')
```
Comment (zhouzaida) [inline setup.py:333]:
```suggestion
                            raise ImportError('failed to download mlu-ops')
```

Comment (DanieeelLiu) [reply setup.py]: Fixed

**Review by zhouzaida — APPROVED** [review]

File: setup.py [hunk 7]
```diff
@@ -300,9 +300,37 @@ def get_extensions():
...
+                with open(file_path) as f:
+                    for line in f:
+                        if re.search('MLUOP_MAJOR', line):
+                            major = line.strip().split(' ')[2]
+                        if re.search('MLUOP_MINOR', line):
+                            minor = line.strip().split(' ')[2]
+                        if re.search('MLUOP_PATCHLEVEL', line):
+                            patchlevel = line.strip().split(' ')[2]
+                f.close()
+                version_str = 'v$major.$minor.$patchlevel'
```
Comment (Qiza-lyhm) [inline setup.py:318]: 这个语法错了吧，这不是shell是python，应该是version_str = f'v{major}.{minor}.{patchlevel}'

**Review by WangYuan-cn — APPROVED** [review]

File: setup.py [hunk 8]
```diff
@@ -299,17 +300,105 @@ def get_extensions():
...
+                    raise FileExistsError(
+                        'mlu-ops already exists, please move it out,'
+                        'or rename or remove it.')
+            else:
+                if not os.path.exists('mlu-ops'):
+                    import requests
+                    mluops_url = 'https://github.com/Cambricon/mlu-ops/' + \
+                        'archive/refs/tags/' + mmcv_mluops_version + '.zip'
+                    req = requests.get(mluops_url)
+                    with open('./mlu-ops-latest.zip', 'wb') as f:
```
Comment (defei-coder) [inline setup.py:353]: Still use this file name?

File: setup.py [hunk 9]
```diff
@@ -299,17 +300,105 @@ def get_extensions():
...
+                        try:
+                            f.write(req.content)
+                        except Exception:
+                            raise ImportError('failed to download mlu-ops')
+
+                    from zipfile import BadZipFile, ZipFile
+                    with ZipFile('./mlu-ops-latest.zip', 'r') as archive:
```
Comment (defei-coder) [inline setup.py:360]: Still use this file name?

File: setup.py [hunk 10]
```diff
@@ -299,17 +300,105 @@ def get_extensions():
...
+                        try:
+                            archive.extractall()
+                            dir_name = archive.namelist()[0].split('/')[0]
+                            os.rename(dir_name, 'mlu-ops')
+                        except BadZipFile:
+                            print('invalid mlu-ops-latest.zip file')
```
Comment (defei-coder) [inline setup.py:366]: Still use this file name?

File: setup.py (hunk 8 above)
Comment (DanieeelLiu) [reply setup.py]: Fixed

File: setup.py (hunk 9 above)
Comment (DanieeelLiu) [reply setup.py]: Fixed

File: setup.py (hunk 10 above)
Comment (DanieeelLiu) [reply setup.py]: Fixed

File: setup.py [hunk 11]
```diff
@@ -299,17 +300,102 @@ def get_extensions():
...
+            def get_mluops_version(file_path):
+                with open(file_path) as f:
+                    for line in f:
+                        if re.search('MLUOP_MAJOR', line):
+                            major = line.strip().split(' ')[2]
+                        if re.search('MLUOP_MINOR', line):
+                            minor = line.strip().split(' ')[2]
+                        if re.search('MLUOP_PATCHLEVEL', line):
+                            patchlevel = line.strip().split(' ')[2]
+                f.close()
```
Comment (zhouzaida) [inline setup.py:316]:
```suggestion
```

File: setup.py [hunk 12]
```diff
@@ -299,17 +300,102 @@ def get_extensions():
         elif (hasattr(torch, 'is_mlu_available') and
                 torch.is_mlu_available()) or \
                 os.getenv('FORCE_MLU', '0') == '1':
+            import re
+
```
Comment (zhouzaida) [inline setup.py:304]: `re` has been imported in the head of the file.

File: setup.py (hunk 11 above)
Comment (DanieeelLiu) [reply setup.py]: Fixed

File: setup.py (hunk 12 above)
Comment (DanieeelLiu) [reply setup.py]: Fixed

**Review by zhouzaida — APPROVED** [review]

**Pull request merged**